### ISSUE TITLE
Request for Guidance on Converting Transformer-405m JAX Weights to PyTorch for LM Evaluation

### ISSUE THREAD

**Issue by szrrr04** [Caiyun-AI/DCFormer#21]
Hello! I made some improvements on your DCFormer-405m model and performed pre-training and evaluation. Now, I plan to migrate these improvements to the Transformer-405m in your repository to see the results. The pre-training has been successfully done, but I'm still unsure about how to convert the pre- and post-improvement Transformer JAX models to PyTorch and use the lm-evaluation library for downstream task evaluation. It seems that the pytorch folder in the repository currently only contains files for the DCFormer conversion? If I want to convert the pre- and post-improvement Transformer-JAX training weights, should I modify the files in the DCFormer folder under pytorch to make them compatible with Transformer? Or is there a simpler way? For example, would it be possible for you to provide the script or file that you used to convert the Transformer-trained weights from JAX to PyTorch? I would really appreciate it!

**Comment by hilbertmeng** [comment]
We provide [an example](https://github.com/Caiyun-AI/DCFormer/blob/main/pytorch/dcformer/modeling_dcformer.py#L805-L849) to convert DCFormer from JAX to PyTorch. The script to convert weights depends on your own model implementation, so we can only provide a general conversion procedure:
1. Implement the pre- and post-improvement Transformer models in both JAX and PyTorch.
2. List all weights in both models and match them. (Mind the Linear dimension order in JAX and PyTorch.)
3. Validate the conversion by comparing their loss on a small valid dataset(eg. eight 2k sequences) with difference commonly less than or around 0.001 empirically.

When there is a bug during model conversion, you can print statistics(eg. variance, l2-norm) of every layer's activations and all intermediate computation results to debug. Hope it may help.
